pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeSegmentDataManager.java

```diff
@@ -1725,6 +1723,19 @@ public RealtimeSegmentDataManager(SegmentZKMetadata segmentZKMetadata, TableConf
     }
   }
 
+  // Consumption while downloading and replacing the slow replicas is not allowed for the following use cases:
+  // 1. Partial Upserts
+  // 2. Dedup Tables
+  // For the above table types, we would be looking into the metadata information when inserting a new record,
+  // so it is not ideal to allow consumption when downloading and replacing the consuming segment as we might see
+  // unusual behaviour of duplicates in dedup tables and inconsistent entries compared to lead replicas for partial
+  // upsert tables.
+  private boolean isConsumptionAllowedDuringCommit() {
+    return (!_realtimeTableDataManager.isPartialUpsertEnabled() && !_realtimeTableDataManager.isDedupEnabled()) || (
+        _tableConfig.getUpsertConfig() != null && _tableConfig.getUpsertConfig()
+            .isAllowPartialUpsertConsumptionDuringCommit());
```

Comment:
maybe this? so it's more clear that isAllowPartialUpsertConsumptionDuringCommit is only used for upsert table. 
```
return !_realtimeTableDataManager.isDedupEnabled() && (!_realtimeTableDataManager.isPartialUpsertEnabled() || _tableConfig.getUpsertConfig().isAllowPartialUpsertConsumptionDuringCommit())
```